src/App.jsx

```diff
@@ -1,22 +1,78 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useReducer } from 'react';
 import './App.css';
 import TodoList from './TodoList';
 import AddTodoForm from './AddTodoForm';
 
-function useSemiPersistentState() {
+const todoListReducer = (state, action) => {
+	switch (action.type) {
+		case 'FETCH_INIT':
+			return {
+				...state,
+				isLoading: true,
+				isError: false
+			};
+		case 'FETCH_SUCCESS':
+			return {
+				...state,
+				isLoading: false,
+				isError: false,
+				data: action.payload
+			};
+		case 'FETCH_FAILURE':
+			return {
+				...state,
+				isLoading: false,
+				isError: true
+			};
+		case 'ADD':
+			return {
+				...state,
+				data: [...state.data, action.payload]
+			};
+		case 'REMOVE':
+			return {
+				...state,
+				data: state.data.filter(todo => todo.id !== action.payload.id)
+			};
+		default:
+			throw new Error('Unsupported todoList action type.');
+	}
+};
```

Comment:
Not part of the assignment instructions but I am a fan of a reducer approach to controlling multiple state values. It's a bit verbose but so much easier to consolidate business logic than independent handler functions.